diff --git a/Makefile b/Makefile
index cf21cb8..56f6693 100644
--- a/Makefile
+++ b/Makefile
@@ -6,6 +6,10 @@
 # Uncomment exactly one of the lines labelled (A), (B), and (C) below
 # to switch between compilation modes.
 
+PREFIX ?= /usr/local
+LIBDIR ?= lib
+INCLUDEDIR ?= include
+
 # (A) Production use (optimized mode)
 OPT ?= -O2 -DNDEBUG
 # (B) Debug mode, w/ full line-level debugging symbols
@@ -102,7 +106,21 @@ libmemenv.$(SHARED3): $(MEMENV_SOURCES)
 
 endif  # PLATFORM_SHARED_EXT
 
-all: libleveldb.SHARED $(LIBRARY)
+all: libleveldb.SHARED libmemenv.SHARED $(LIBRARY)
+
+install: all
+	install -d $(DESTDIR)$(PREFIX)/$(LIBDIR)
+	install -d $(DESTDIR)$(PREFIX)/$(INCLUDEDIR)/leveldb
+	echo "end string"
+ifneq ($(PLATFORM_SHARED_EXT),)
+	install -m 0644 libleveldb.$(SHARED3) $(DESTDIR)$(PREFIX)/$(LIBDIR)
+	install -m 0644 libmemenv.$(SHARED3) $(DESTDIR)$(PREFIX)/$(LIBDIR)
+	cd $(DESTDIR)$(PREFIX)/$(LIBDIR) && ln -sf libleveldb.$(SHARED3) libleveldb.$(SHARED2)
+	cd $(DESTDIR)$(PREFIX)/$(LIBDIR) && ln -sf libleveldb.$(SHARED3) libleveldb.$(SHARED1)
+	cd $(DESTDIR)$(PREFIX)/$(LIBDIR) && ln -sf libmemenv..$(SHARED3) libmemenv.$(SHARED2)
+	cd $(DESTDIR)$(PREFIX)/$(LIBDIR) && ln -sf libmemenv.$(SHARED3) libmemenv.$(SHARED1)
+endif
+	install -m 0644 include/leveldb/* $(DESTDIR)$(PREFIX)/$(INCLUDEDIR)/leveldb
 
 check: all $(PROGRAMS) $(TESTS)
 	for t in $(TESTS); do echo "***** Running $$t"; ./$$t || exit 1; done
